src/server/dcp_server/models.py

```diff
@@ -60,12 +75,475 @@ def masks_to_outlines(self, mask):
+        return final_mask
+
+    def find_max_patch_size(self, mask):
+
+        # Find objects in the mask
+        objects = find_objects(mask)
+
+        # Initialize variables to store the maximum patch size
+        max_patch_size = 0
+        max_patch_indices = None
+
+        # Iterate over the found objects
+        for obj in objects:
+            # Extract start and stop values from the slice object
+            slices = [s for s in obj]
+            start = [s.start for s in slices]
+            stop = [s.stop for s in slices]
+
+            # Calculate the size of the patch along each axis
+            patch_size = tuple(stop[i] - start[i] for i in range(len(start)))
+
+            # Calculate the total size (area) of the patch
+            total_size = 1
+            for size in patch_size:
+                total_size *= size
+
+            # Check if the current patch size is larger than the maximum
+            if total_size > max_patch_size:
+                max_patch_size = total_size
+                max_patch_indices = obj
+            
+            max_patch_size_edge = np.ceil(np.sqrt(max_patch_size))
+
+            return max_patch_size_edge
+
+    def pad_centered_padded_patch(self, x: np.ndarray, c, p, mask: np.ndarray=None, noise_intensity=None) -> np.ndarray:
+        """
+        Crop a patch from an array `x` centered at coordinates `c` with size `p`, and apply padding if necessary.
+
+        Args:
+            x (np.ndarray): The input array from which the patch will be cropped.
+            c (tuple): The coordinates (row, column, channel) at the center of the patch.
+            p (tuple): The size of the patch to be cropped (height, width).
+
+        Returns:
+            np.ndarray: The cropped patch with applied padding.
+        """
+        if mask.shape[0] < mask.shape[-1]:
+            if isinstance(mask,torch.Tensor):
+                mask = mask.permute(1,2,0).numpy()
+            else:
+                mask = mask.transpose(1, 2, 0)
+
+        height, width = p  # Size of the patch
+
+        # Calculate the boundaries of the patch
+        top = c[0] - height // 2
+        bottom = top + height
+        
+        left = c[1] - width // 2
+        right = left + width
+
+        # Crop the patch from the input array
+
+        if mask is not None:
+
+            mask_ = mask.max(-1) if len(mask.shape) >= 3 else mask
+            # central_label = mask_[c[0], c[1]]
+            central_label = mask_[c[0]][c[1]]
+            # Zero out values in the patch where the mask is not equal to the central label
+       
+            # m = (mask_ != central_label) & (mask_ > 0)
+            m = (mask_ != central_label) & (mask_ > 0)
+            x[m] = 0
+
+            if noise_intensity is not None:
+                x[m] = np.random.normal(scale=noise_intensity, size=x[m].shape)
+
+        patch = x[max(top, 0):min(bottom, x.shape[0]), max(left, 0):min(right, x.shape[1])]
+        
+        if len(c) == 3:
```

Comment:
we only call the crop_centered_padded_patch function in get_centered_patches, where c is passed as (cx, cy). Hence c can never have len=3. That's why I suggest to remove this if statement

Reply:
"get_centered_patches" is used in both the train and eval functions.
In the train function, it's a part of the create_patch_dataset function, which is used during training. The code looks like this:
```
patches, labels = self.create_patch_dataset(imgs, masks_classes, masks_instances)
```
In this case, 'imgs' are 3-channel images, where the last dimension represents the number of channels.

In the eval function, the code is as follows:
```
patches, _ = self.get_centered_patches(img, instance_mask, self.eval_config["classifier"]["data"]["patch_size"], noise_intensity=5)
```
In this case, 'img' is an image with dimensions HxW.

Reply:
get_centered patches is used in both train and eval functions yes. However `if len(c) == 3:` is written in `crop_centered_padded_patch` which is only called once like this:
```
patch = self.crop_centered_padded_patch(img.copy(),
                                       (c_x, c_y),
                                       (p_size, p_size),
                                       mask=mask,
                                       noise_intensity=noise_intensity)
```
this is why c can never have len=3, since c is only once given as (cx,cy). I'm removing this for now to avoid confusion